pkg/util/sqlexec/restricted_sql_executor.go

```diff
@@ -199,6 +199,14 @@ type RecordSet interface {
 	Close() error
 }
 
+// DetachableRecordSet extends the `RecordSet` to support detaching from current session context
+type DetachableRecordSet interface {
+	RecordSet
+
+	// Detach detaches the record set from the current session context.
+	Detach() (RecordSet, error)
```

Comment:
Could we make this function to `TryDetach() (RecordSet, bool)` because and recordSet that cannot be detach may be a normal case

Reply:
I'll return `TryDetach() (RecordSet, bool, error)` to indicate whether the statement is suitable to detach, and whether it detaches successfully. I sounds strange but I didn't come up a better way :facepalm:.

If the caller receives `nil, false, _`, it means the original record set can still be used. If the caller receives `_, _, err`, it means the state of session/recordSet is in a chaos, and usually the connection will be halted.